packages/tools/src/index.ts

```diff
@@ -0,0 +1,179 @@
+// @ts-ignore
+import AsBind from "as-bind/dist/as-bind.cjs.js";
+import fs from "fs";
+
+type ToolParameters = {
+  type: string | "object";
+  properties: Record<string, { type: string; description?: string }>;
+  required?: string[];
+};
+
+interface ToolMetadata {
+  description: string;
+  name: string;
+  parameters?: ToolParameters;
+  argsKwargs?: Record<string, any>;
+}
+
+interface BaseTool {
+  call?: (...args: any[]) => any;
+  metadata: ToolMetadata;
+}
+
+export default class ToolFactory {
+  /**
+   * Transform the metadata from the assemblyscript raw format to the application format
+   * Convert asm string to ts string
+   * Convert asm array to ts array
+   * Convert the properties from an Array to a Record<string, { type: string; description?: string }>
+   * Convert the argsKwargs from an Array to a Record<string, any>
+   */
+  private static wasmInstanceToConfigs(wasmInstance: any): BaseTool {
+    const { __pin, __unpin, __getString, __getArray, __newString } =
+      wasmInstance.exports;
+
+    const getObjectByAddress = (address: any, classWrapper: any) => {
```

Comment:
can we extract her reusable transform operations (e.g. Map to Record)? I guess this would make the code easier to read